orchagent/aclorch.cpp

```diff
@@ -2142,6 +2221,100 @@ void AclRuleMirror::onUpdate(SubjectType type, void *cntx)
     }
 }
 
+AclRuleUnderlaySetDhcp::AclRuleUnderlaySetDhcp(AclOrch *aclOrch, string rule, string table, MetaDataMgr* m_metaDataMgr, bool createCounter):
+    AclRule(aclOrch, rule, table, createCounter),
+    table_id(table),
+    m_metaDataMgr(m_metaDataMgr)
+{
+}
+
+uint32_t AclRuleUnderlaySetDhcp::getDscpValue() const
+{
+    return cachedDscpValue;
+}
+
+uint32_t AclRuleUnderlaySetDhcp::getMetadata() const
+{
+    return cachedMetadata;
+}
+
+bool AclRuleUnderlaySetDhcp::validateAddAction(string attr_name, string _attr_value)
+{
+    SWSS_LOG_ENTER();
+
+    string attr_value = to_upper(_attr_value);
+
+    sai_object_id_t table_oid = m_pAclOrch->getTableById(table_id);
+    auto aclTable = m_pAclOrch->getTableByOid(table_oid);
+    string type = aclTable->type.getName();
+    string key = table_id + ":" + m_id;
+    // we handle the allocation of metadata for here. based on SET_DSCP action, we check if a metadata is already allocated then we reuse it
+    // otherwise we allocate a new metadata. This metadata is then set an the action for the Rule of this table. We also cache the SET_DSCP
+    // value and the allocated metadata in a the rule structure itself so that when we go to addRule we can use these to add the
+    // egr_set_dscp rule
+    if (attr_name == ACTION_DSCP && (type == TABLE_TYPE_MARK_META || type == TABLE_TYPE_MARK_META_V6))
+    {
+        if (!m_pAclOrch->isUsingEgrSetDscp(table_id))
+        {
+
+            SWSS_LOG_ERROR("Unexpected Error. Table %s not asssociated with EGR_SET_DSCP table", table_id.c_str());
+            return false;
+        }
+
+        u_int8_t actionDscpValue = uint8_t(std::stoi(attr_value));
+        cachedDscpValue = actionDscpValue;
+        auto metadata = m_metaDataMgr->getFreeMetaData(actionDscpValue);
+
+        if (metadata == METADATA_VALUE_INVALID)
+        {
+            SWSS_LOG_ERROR("Failed to get free metadata for DSCP value %d", actionDscpValue);
+            return false;
+        }
+        cachedMetadata = metadata;
+        attr_name = ACTION_META_DATA;
```

Comment:
Why the input `attr_name` is changed here?

Reply:
The table configured is **UNDERLAY_SET_DSCP/V6** with **ACTION SET_DSCP**. When a new entry comes in it has **ACTION_SET_DSCP** with a dscp value. Here the action is changed to **SET_METADATA** and the allocated metadata value is set with this action for this entry. The dscp value is stored in the **cachedDscpValue** which is an internal variable of the **AclRuleUnderlaySetDscp** structure. Before this entry is installed in **addAclRule**, a **EGR_SET_DSCP** rule is created and the cachedDscpValue is set as its action while the cachedMetadata allocated here is used as its match criteria.